File: conformance/runners/run_invariants.py
# SPDX-License-Identifier: Apache-2.0
#
# TLA+-invariant cross-language conformance runner — Python lane (SCAFFOLD).
#
# STATUS: SCAFFOLDED, NOT WIRED. This lane parses the shared invariant corpus
# (conformance/invariants.json) and dispatches every case by domain, exactly as
# the JavaScript lane (conformance/runners/run-invariants.mjs) does, so that when
# the state machines gain Python ports this file drives them with zero corpus
# changes and any cross-port divergence surfaces immediately.
#
# WIRING BLOCKER (precise): the invariants are STATE-MACHINE properties, not
# receipt-verification properties. Driving them in Python requires a Python port
# of the two production state machines the JS lane drives:
#   * capability domain -> createMemoryCapabilityStore (register/reserveSpend/
#                          commitSpend) in packages/gate/capability-receipt.js
#   * handshake domain  -> checkNoDuplicateResult / checkResultImmutability /
#                          checkNotExpired / checkBindingValid in
#                          lib/handshake/invariants.js
# packages/python-verify/emilia_verify today verifies RECEIPT vectors only; it
# has NO capability store and NO handshake invariant module. Until such a port
# exists, wiring this lane would mean re-implementing the state machines here,
# which is a NEW implementation the author writes — it cannot honestly detect a
# divergence from itself. So this lane is intentionally left unwired and reports
# each case as SKIPPED(unwired) rather than fabricating a pass.
#
# When a Python port lands (proposed: packages/python-verify/emilia_verify/
# capability_store.py and .../handshake_invariants.py), replace the dispatch
# bodies below and flip UNWIRED = False.
#
#   python3 conformance/runners/run_invariants.py [path/to/invariants.json]
#
# Exit codes: 0 = every case executed and held; 2 = lane is unwired (default
# today) so it can never be mistaken for a passing conformance lane.

import json
import os
import sys

UNWIRED = True  # flip to False once the Python state-machine ports exist.

HERE = os.path.dirname(os.path.abspath(__file__))
DEFAULT_CORPUS = os.path.join(HERE, "..", "invariants.json")


def run_capability_case(case):
    # TODO(python-port): drive a Python createMemoryCapabilityStore equivalent.
    raise NotImplementedError("capability store has no Python port")


def run_handshake_case(case):
    # TODO(python-port): drive Python ports of the handshake invariant guards.
    raise NotImplementedError("handshake invariants have no Python port")


def main():
    argv = [a for a in sys.argv[1:] if not a.startswith("--")]
    corpus_path = argv[0] if argv else DEFAULT_CORPUS
    with open(corpus_path, "r", encoding="utf-8") as fh:
        corpus = json.load(fh)

    total = sum(len(inv["cases"]) for inv in corpus["invariants"])
    if UNWIRED:
        print("EP invariant conformance — Python lane: SCAFFOLD (UNWIRED)")
        print(f"  parsed {total} cases across {len(corpus['invariants'])} invariants from {corpus_path}")
        print("  no Python port of the capability store / handshake invariants exists yet;")
        print("  see the wiring blocker at the top of this file. Reporting SKIPPED(unwired).")
        sys.exit(2)

    failures = 0
    for inv in corpus["invariants"]:
        for case in inv["cases"]:
            cid = f"{inv['invariant']}/{case['name']}"
            try:
                if inv["domain"] == "capability":
                    run_capability_case(case)
                elif inv["domain"] == "handshake":
                    run_handshake_case(case)
                else:
                    raise ValueError(f"unknown domain: {inv['domain']}")
                print(f"  ok   {cid}  {inv['spec']}")
            except Exception as err:  # noqa: BLE001 — report every divergence
                failures += 1
                print(f" FAIL  {cid}  {inv['spec']}\n        -> {err}")
    print(f"\n{total - failures}/{total} invariant cases hold.")
    sys.exit(1 if failures else 0)


if __name__ == "__main__":
    main()

File: conformance/runners/run_invariants.go
// SPDX-License-Identifier: Apache-2.0
//
// TLA+-invariant cross-language conformance runner — Go lane (SCAFFOLD).
//
// STATUS: SCAFFOLDED, NOT WIRED. This lane parses the shared invariant corpus
// (conformance/invariants.json) and walks every case by domain, mirroring the
// JavaScript lane (conformance/runners/run-invariants.mjs), so that once the
// state machines gain Go ports this file drives them with no corpus changes and
// any cross-port divergence surfaces immediately.
//
// WIRING BLOCKER (precise): the invariants are STATE-MACHINE properties. Driving
// them in Go requires a Go port of the two production state machines the JS lane
// drives:
//   - capability domain: createMemoryCapabilityStore (register/reserveSpend/
//     commitSpend) in packages/gate/capability-receipt.js
//   - handshake domain:  checkNoDuplicateResult / checkResultImmutability /
//     checkNotExpired / checkBindingValid in lib/handshake/invariants.js
//
// packages/go-verify today verifies RECEIPT vectors only (trust_receipt.go,
// signoff.go, ...); it has NO capability store and NO handshake invariant
// module. Until such a port exists, wiring this lane would mean re-implementing
// the state machines here — a NEW implementation the author writes, which cannot
// honestly detect a divergence from itself. So this lane is intentionally left
// unwired and reports SKIPPED(unwired) rather than fabricating a pass.
//
// When a Go port lands (proposed: packages/go-verify/capability_store.go and
// packages/go-verify/handshake_invariants.go), replace the dispatch bodies below
// and flip unwired to false.
//
// Standalone (stdlib only): go run conformance/runners/run_invariants.go [corpus.json]
// Exit codes: 0 = every case executed and held; 2 = lane is unwired (default
// today) so it can never be mistaken for a passing conformance lane.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const unwired = true // flip to false once the Go state-machine ports exist.

type invariant struct {
	Invariant string    `json:"invariant"`
	Spec      string    `json:"spec"`
	Domain    string    `json:"domain"`
	Cases     []invCase `json:"cases"`
}

type invCase struct {
	Name    string            `json:"name"`
	Actions []json.RawMessage `json:"actions"`
}

type corpus struct {
	Invariants []invariant `json:"invariants"`
}

func main() {
	args := []string{}
	for _, a := range os.Args[1:] {
		if !strings.HasPrefix(a, "--") {
			args = append(args, a)
		}
	}
	corpusPath := ""
	if len(args) > 0 {
		corpusPath = args[0]
	} else {
		exe, _ := os.Getwd()
		corpusPath = filepath.Join(exe, "conformance", "invariants.json")
	}

	raw, err := os.ReadFile(corpusPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "cannot read corpus %s: %v\n", corpusPath, err)
		os.Exit(2)
	}
	var c corpus
	if err := json.Unmarshal(raw, &c); err != nil {
		fmt.Fprintf(os.Stderr, "cannot parse corpus: %v\n", err)
		os.Exit(2)
	}

	total := 0
	for _, inv := range c.Invariants {
		total += len(inv.Cases)
	}

	if unwired {
		fmt.Println("EP invariant conformance — Go lane: SCAFFOLD (UNWIRED)")
		fmt.Printf("  parsed %d cases across %d invariants from %s\n", total, len(c.Invariants), corpusPath)
		fmt.Println("  no Go port of the capability store / handshake invariants exists yet;")
		fmt.Println("  see the wiring blocker at the top of this file. Reporting SKIPPED(unwired).")
		os.Exit(2)
	}

	// TODO(go-port): dispatch each case by inv.Domain against the Go state-machine
	// ports and compare observed outcomes to the declared expectations.
	failures := 0
	for _, inv := range c.Invariants {
		for _, kase := range inv.Cases {
			_ = kase
			fmt.Printf("  ok   %s/%s  %s\n", inv.Invariant, kase.Name, inv.Spec)
		}
	}
	fmt.Printf("\n%d/%d invariant cases hold.\n", total-failures, total)
	if failures > 0 {
		os.Exit(1)
	}
}
